codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/EndpointHostPrefixMiddleware.java

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ *
+ */
+
+package software.amazon.smithy.go.codegen.integration;
+
+import java.util.ArrayList;
+import java.util.List;
+import software.amazon.smithy.codegen.core.SymbolProvider;
+import software.amazon.smithy.go.codegen.GoDelegator;
+import software.amazon.smithy.go.codegen.GoSettings;
+import software.amazon.smithy.go.codegen.GoStackStepMiddlewareGenerator;
+import software.amazon.smithy.go.codegen.GoWriter;
+import software.amazon.smithy.go.codegen.SmithyGoDependency;
+import software.amazon.smithy.go.codegen.SymbolUtils;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.pattern.SmithyPattern;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.traits.EndpointTrait;
+
+public class EndpointHostPrefixMiddleware implements GoIntegration {
+
+    private static final String MIDDLEWARE_ID = "EndpointHostPrefix";
+
+    List<RuntimeClientPlugin> runtimeClientPlugins = new ArrayList<>();
+    List<OperationShape> endpointPrefixOperations = new ArrayList<>();
+
+    @Override
+    public void processFinalizedModel(GoSettings settings, Model model) {
+        ServiceShape service = settings.getService(model);
+        endpointPrefixOperations = getOperationsWithEndpointPrefix(model, service);
+
+        endpointPrefixOperations.forEach((operation) -> {
+            String middlewareHelperName = getMiddlewareHelperName(operation);
+            runtimeClientPlugins.add(RuntimeClientPlugin.builder()
+                    .operationPredicate((m, s, o) -> o.equals(operation))
```

Comment:
Missed this before, won't this operationPredicate always be true?

Reply:
If I am looking at this correctly, may be having `endpointPrefixOperations` as a Set, rather than a List would be better here.  Then you don't have to iterate over the list.